wandb/integration/metaflow/metaflow.py

```diff
@@ -0,0 +1,272 @@
+
+    if others:
+        artifact = wandb.Artifact(name, type="other")
+        with artifact.new_file(f"{name}.pkl", "wb") as f:
+            pickle.dump(data, f)
+        run.log_artifact(artifact)
+        print(f"wandb: logging artifact: {name} ({type(data)})")
+
+
+@typedispatch
+def wandb_use(name: str, data, *args, **kwargs):
+    try:
+        return _wandb_use(name, data, *args, **kwargs)
+    except wandb.CommError:
+        print(
+            f"This artifact ({name}, {type(data)}) does not exist in the wandb datastore!"
+            f"If you created an instance inline (e.g. sklearn.ensemble.RandomForestClassifier), then you can safely ignore this"
+            f"Otherwise you may want to check your internet connection!"
+        )
+
+
+@typedispatch  # noqa: F811
+def wandb_use(name: str, data: (dict, list, set, str, int, float, bool), *args, **kwargs):  # type: ignore
+    pass  # do nothing for these types
+
+
+@typedispatch  # noqa: F811
+def _wandb_use(name: str, data: (nn.Module, BaseEstimator), models=False, run=None, testing=False, *args, **kwargs):  # type: ignore
+    if testing:
+        return "models" if models else None
+
+    if models:
+        run.use_artifact(f"{name}:latest")
+        print(f"wandb: using artifact: {name} ({type(data)})")
+
+
+@typedispatch  # noqa: F811
+def _wandb_use(name: str, data: (pd.DataFrame, Path), datasets=False, run=None, testing=False, *args, **kwargs):  # type: ignore
+    if testing:
+        return "datasets" if datasets else None
+
+    if datasets:
+        run.use_artifact(f"{name}:latest")
+        print(f"wandb: using artifact: {name} ({type(data)})")
+
+
+@typedispatch  # noqa: F811
+def _wandb_use(name: str, data, others=False, run=None, testing=False, *args, **kwargs):  # type: ignore
+    if testing:
+        return "others" if others else None
+
+    if others:
+        run.use_artifact(f"{name}:latest")
+        print(f"wandb: using artifact: {name} ({type(data)})")
+
+
+def wandb_log(
+    func=None,
+    # /,  # py38 only
+    datasets=False,
+    models=False,
+    others=False,
+    settings=None,
+):
+    @wraps(func)
+    def decorator(func):
+        # If you decorate a class, apply the decoration to all methods in that class
+        if inspect.isclass(func):
+            cls = func
+            for attr in cls.__dict__:
+                if callable(getattr(cls, attr)):
+                    if not hasattr(attr, "_base_func"):
+                        setattr(cls, attr, decorator(getattr(cls, attr)))
+            return cls
+
+        # prefer the earliest decoration (i.e. method decoration overrides class decoration)
+        if hasattr(func, "_base_func"):
+            return func
+
+        @wraps(func)
+        def wrapper(self, *args, settings=settings, **kwargs):
+            if not settings:
+                settings = wandb.Settings(
+                    run_group=f"{current.flow_name}/{current.run_id}",
+                    run_job_type=current.step_name,
```

Comment:
I think we should set these even when settings are specified if the user didn't override them.

Reply:
done